File: libs/mdblist.php
<?php
/**
 * MDBList Integration Library
 * Fetch movie/TV lists from MDBList.com for playlist automation
 * 
 * MDBList provides curated lists with TMDB IDs, IMDB IDs, and metadata
 * which can be directly integrated into the playlist generation pipeline.
 */

class MDBListProvider {
    private $apiKey;
    private $cacheDir;
    private $cacheExpiry = 3600; // 1 hour cache
    
    public function __construct($apiKey = null) {
        $this->apiKey = $apiKey ?: ($GLOBALS['MDBLIST_API_KEY'] ?? '');
        $this->cacheDir = __DIR__ . '/../cache/mdblist/';
        
        if (!is_dir($this->cacheDir)) {
            @mkdir($this->cacheDir, 0755, true);
        }
    }
    
    /**
     * Set the API key
     */
    public function setApiKey($apiKey) {
        $this->apiKey = $apiKey;
    }
    
    /**
     * Check if API key is configured
     */
    public function hasApiKey() {
        return !empty($this->apiKey);
    }
    
    /**
     * Test API connection
     */
    public function testConnection() {
        if (!$this->hasApiKey()) {
            return ['success' => false, 'error' => 'API key not configured'];
        }
        
        $url = "https://mdblist.com/api/user/?apikey=" . urlencode($this->apiKey);
        $response = $this->fetchUrl($url);
        
        if ($response === false) {
            return ['success' => false, 'error' => 'Failed to connect to MDBList API'];
        }
        
        $data = json_decode($response, true);
        if (isset($data['error'])) {
            return ['success' => false, 'error' => $data['error']];
        }
        
        return [
            'success' => true,
            'user' => $data['name'] ?? 'Unknown',
            'api_requests_remaining' => $data['api_requests'] ?? 'Unknown'
        ];
    }
    
    /**
     * Get user's MDBList lists
     */
    public function getUserLists() {
        if (!$this->hasApiKey()) {
            return ['success' => false, 'error' => 'API key not configured'];
        }
        
        $url = "https://mdblist.com/api/lists/user/?apikey=" . urlencode($this->apiKey);
        $response = $this->fetchUrl($url);
        
        if ($response === false) {
            return ['success' => false, 'error' => 'Failed to fetch user lists'];
        }
        
        $data = json_decode($response, true);
        if (isset($data['error'])) {
            return ['success' => false, 'error' => $data['error']];
        }
        
        return ['success' => true, 'lists' => $data];
    }
    
    /**
     * Fetch items from a public MDBList URL
     * Supports URLs like: https://mdblist.com/lists/username/listname
     */
    public function fetchListByUrl($url) {
        // Parse the URL to extract username and list name
        if (preg_match('/mdblist\.com\/lists\/([^\/]+)\/([^\/\?]+)/', $url, $matches)) {
            $username = $matches[1];
            $listSlug = $matches[2];
            return $this->fetchPublicList($username, $listSlug);
        }
        
        // Try as a list ID
        if (preg_match('/mdblist\.com\/lists\/(\d+)/', $url, $matches)) {
            return $this->fetchListById($matches[1]);
        }
        
        return ['success' => false, 'error' => 'Invalid MDBList URL format'];
    }
    
    /**
     * Fetch a public list by username and slug
     */
    public function fetchPublicList($username, $listSlug) {
        // Check cache first
        $cacheKey = "list_{$username}_{$listSlug}";
        $cached = $this->getCache($cacheKey);
        if ($cached !== null) {
            return $cached;
        }
        
        // MDBList public list JSON endpoint
        $url = "https://mdblist.com/lists/{$username}/{$listSlug}/json";
        $response = $this->fetchUrl($url);
        
        if ($response === false) {
            return ['success' => false, 'error' => 'Failed to fetch list'];
        }
        
        $data = json_decode($response, true);
        if ($data === null) {
            return ['success' => false, 'error' => 'Invalid JSON response'];
        }
        
        // Process the list items
        $result = $this->processListItems($data, "{$username}/{$listSlug}");
        $this->setCache($cacheKey, $result);
        
        return $result;
    }
    
    /**
     * Fetch list by ID (requires API key)
     */
    public function fetchListById($listId) {
        if (!$this->hasApiKey()) {
            return ['success' => false, 'error' => 'API key required for list ID lookup'];
        }
        
        // Check cache
        $cacheKey = "list_id_{$listId}";
        $cached = $this->getCache($cacheKey);
        if ($cached !== null) {
            return $cached;
        }
        
        $url = "https://mdblist.com/api/lists/{$listId}/items/?apikey=" . urlencode($this->apiKey);
        $response = $this->fetchUrl($url);
        
        if ($response === false) {
            return ['success' => false, 'error' => 'Failed to fetch list'];
        }
        
        $data = json_decode($response, true);
        if (isset($data['error'])) {
            return ['success' => false, 'error' => $data['error']];
        }
        
        $result = $this->processListItems($data, "list_$listId");
        $this->setCache($cacheKey, $result);
        
        return $result;
    }
    
    /**
     * Search MDBList for lists
     */
    public function searchLists($query) {
        $url = "https://mdblist.com/api/lists/search/?query=" . urlencode($query);
        if ($this->hasApiKey()) {
            $url .= "&apikey=" . urlencode($this->apiKey);
        }
        
        $response = $this->fetchUrl($url);
        if ($response === false) {
            return ['success' => false, 'error' => 'Search failed'];
        }
        
        $data = json_decode($response, true);
        return ['success' => true, 'lists' => $data ?? []];
    }
    
    /**
     * Get top/popular MDBList lists
     */
    public function getTopLists($type = 'movie') {
        $url = "https://mdblist.com/api/lists/top/?mediatype=" . urlencode($type);
        if ($this->hasApiKey()) {
            $url .= "&apikey=" . urlencode($this->apiKey);
        }
        
        $response = $this->fetchUrl($url);
        if ($response === false) {
            return ['success' => false, 'error' => 'Failed to fetch top lists'];
        }
        
        $data = json_decode($response, true);
        return ['success' => true, 'lists' => $data ?? []];
    }
    
    /**
     * Process list items into our standard format
     */
    private function processListItems($items, $sourceName) {
        $movies = [];
        $series = [];
        
        if (!is_array($items)) {
            return [
                'success' => true,
                'source' => $sourceName,
                'movies' => [],
                'series' => [],
                'total' => 0
            ];
        }
        
        foreach ($items as $item) {
            // Skip items without TMDB ID
            $tmdbId = $item['tmdb_id'] ?? $item['id'] ?? null;
            if (!$tmdbId) continue;
            
            $mediaType = $item['mediatype'] ?? $item['media_type'] ?? 'movie';
            
            $processed = [
                'id' => (int)$tmdbId,
                'tmdb_id' => (int)$tmdbId,
                'imdb_id' => $item['imdb_id'] ?? null,
                'title' => $item['title'] ?? $item['name'] ?? 'Unknown',
                'year' => $item['release_year'] ?? $item['year'] ?? null,
                'rating' => $item['score'] ?? $item['rating'] ?? null,
                'poster_path' => $item['poster'] ?? null,
                'backdrop_path' => $item['backdrop'] ?? null,
                'overview' => $item['description'] ?? $item['overview'] ?? '',
                'source' => 'mdblist:' . $sourceName
            ];
            
            if ($mediaType === 'movie') {
                $movies[] = $processed;
            } else {
                // TV show
                $processed['name'] = $processed['title'];
                $series[] = $processed;
            }
        }
        
        return [
            'success' => true,
            'source' => $sourceName,
            'movies' => $movies,
            'series' => $series,
            'total' => count($movies) + count($series),
            'movie_count' => count($movies),
            'series_count' => count($series),
            'fetched_at' => date('Y-m-d H:i:s')
        ];
    }
    
    /**
     * Fetch all configured MDBList lists and merge results
     */
    public function fetchAllConfiguredLists() {
        // Get lists from saved config (mdblist_config.json), not from GLOBALS
        $savedLists = self::getSavedLists();
        $enabledLists = array_filter($savedLists, function($list) {
            return $list['enabled'] ?? true;
        });
        
        if (empty($enabledLists)) {
            return [
                'success' => true,
                'movies' => [],
                'series' => [],
                'total' => 0,
                'message' => 'No MDBList URLs configured or enabled'
            ];
        }
        
        $allMovies = [];
        $allSeries = [];
        $errors = [];
        $seenMovieIds = [];
        $seenSeriesIds = [];
        
        foreach ($enabledLists as $listConfig) {
            $listUrl = $listConfig['url'];
            $result = $this->fetchListByUrl($listUrl);
            
            if (!$result['success']) {
                $errors[] = "Failed to fetch $listUrl: " . ($result['error'] ?? 'Unknown error');
                continue;
            }
            
            // Deduplicate movies
            foreach ($result['movies'] as $movie) {
                if (!isset($seenMovieIds[$movie['id']])) {
                    $seenMovieIds[$movie['id']] = true;
                    $allMovies[] = $movie;
                }
            }
            
            // Deduplicate series
            foreach ($result['series'] as $show) {
                if (!isset($seenSeriesIds[$show['id']])) {
                    $seenSeriesIds[$show['id']] = true;
                    $allSeries[] = $show;
                }
            }
        }
        
        return [
            'success' => true,
            'movies' => $allMovies,
            'series' => $allSeries,
            'total' => count($allMovies) + count($allSeries),
            'movie_count' => count($allMovies),
            'series_count' => count($allSeries),
            'lists_processed' => count($enabledLists),
            'errors' => $errors,
            'fetched_at' => date('Y-m-d H:i:s')
        ];
    }
    
    /**
     * Convert MDBList items to playlist format
     */
    public function toPlaylistFormat($items, $type = 'movie', $playVodUrl = '') {
        $playlist = [];
        $num = 1;
        
        foreach ($items as $item) {
            $entry = [
                'num' => $num++,
                'name' => $item['title'] ?? $item['name'],
                'stream_type' => $type === 'movie' ? 'movie' : 'series',
                'stream_id' => $item['id'],
                'stream_icon' => $item['poster_path'] ?? '',
                'rating' => $item['rating'] ?? 0,
                'added' => time(),
                'category_id' => 'mdblist',
                'container_extension' => 'mp4',
                'custom_sid' => '',
                'direct_source' => ''
            ];
            
            if ($type === 'movie') {
                $entry['year'] = $item['year'] ?? '';
                $entry['plot'] = $item['overview'] ?? '';
            } else {
                $entry['series_id'] = $item['id'];
                $entry['cover'] = $item['poster_path'] ?? '';
            }
            
            $playlist[] = $entry;
        }
        
        return $playlist;
    }
    
    /**
     * HTTP fetch helper
     */
    private function fetchUrl($url, $timeout = 30) {
        $context = stream_context_create([
            'http' => [
                'timeout' => $timeout,
                'header' => [
                    'Accept: application/json',
                    'User-Agent: TMDB-VOD-Playlist/1.0'
                ]
            ],
            'ssl' => [
                'verify_peer' => true,
                'verify_peer_name' => true
            ]
        ]);
        
        $response = @file_get_contents($url, false, $context);
        return $response;
    }
    
    /**
     * Cache helpers
     */
    private function getCacheFile($key) {
        return $this->cacheDir . md5($key) . '.json';
    }
    
    private function getCache($key) {
        $file = $this->getCacheFile($key);
        if (!file_exists($file)) return null;
        
        $data = json_decode(file_get_contents($file), true);
        if (!$data || !isset($data['expires']) || $data['expires'] < time()) {
            @unlink($file);
            return null;
        }
        
        return $data['data'];
    }
    
    private function setCache($key, $data) {
        $file = $this->getCacheFile($key);
        $cacheData = [
            'expires' => time() + $this->cacheExpiry,
            'data' => $data
        ];
        file_put_contents($file, json_encode($cacheData));
    }
    
    /**
     * Clear all MDBList cache
     */
    public function clearCache() {
        $files = glob($this->cacheDir . '*.json');
        $count = 0;
        foreach ($files as $file) {
            if (unlink($file)) $count++;
        }
        return $count;
    }
    
    /**
     * Get saved MDBList URLs from JSON config
     */
    public static function getSavedLists() {
        $configFile = __DIR__ . '/../cache/mdblist_config.json';
        if (!file_exists($configFile)) {
            return [];
        }
        $data = json_decode(file_get_contents($configFile), true);
        return $data['lists'] ?? [];
    }
    
    /**
     * Save MDBList URLs to JSON config
     */
    public static function saveLists($lists) {
        $configFile = __DIR__ . '/../cache/mdblist_config.json';
        $data = [
            'lists' => $lists,
            'updated' => date('Y-m-d H:i:s')
        ];
        return file_put_contents($configFile, json_encode($data, JSON_PRETTY_PRINT)) !== false;
    }
    
    /**
     * Add a list URL
     */
    public static function addList($url, $name = '', $enabled = true) {
        $lists = self::getSavedLists();
        
        // Check if already exists
        foreach ($lists as $list) {
            if ($list['url'] === $url) {
                return ['success' => false, 'error' => 'List already exists'];
            }
        }
        
        $lists[] = [
            'url' => $url,
            'name' => $name ?: self::extractListName($url),
            'enabled' => $enabled,
            'added' => date('Y-m-d H:i:s')
        ];
        
        self::saveLists($lists);
        return ['success' => true];
    }
    
    /**
     * Remove a list URL
     */
    public static function removeList($url) {
        $lists = self::getSavedLists();
        $lists = array_filter($lists, function($list) use ($url) {
            return $list['url'] !== $url;
        });
        self::saveLists(array_values($lists));
        return ['success' => true];
    }
    
    /**
     * Toggle list enabled status
     */
    public static function toggleList($url, $enabled) {
        $lists = self::getSavedLists();
        foreach ($lists as &$list) {
            if ($list['url'] === $url) {
                $list['enabled'] = $enabled;
                break;
            }
        }
        self::saveLists($lists);
        return ['success' => true];
    }
    
    /**
     * Extract list name from URL
     */
    private static function extractListName($url) {
        if (preg_match('/mdblist\.com\/lists\/([^\/]+)\/([^\/\?]+)/', $url, $matches)) {
            return ucwords(str_replace('-', ' ', $matches[2]));
        }
        return 'MDBList';
    }
}

File: internal/services/mdblist.go
package services

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"time"
)

// MDBListClient handles MDBList.com API integration
type MDBListClient struct {
	apiKey     string
	cacheDir   string
	client     *http.Client
	cacheExpiry time.Duration
}

// MDBListItem represents a media item from MDBList
type MDBListItem struct {
	ID           int     `json:"id"`
	TMDBID       int     `json:"tmdb_id"`
	IMDBID       string  `json:"imdb_id,omitempty"`
	Title        string  `json:"title"`
	Year         int     `json:"year,omitempty"`
	Rating       float64 `json:"rating,omitempty"`
	PosterPath   string  `json:"poster_path,omitempty"`
	BackdropPath string  `json:"backdrop_path,omitempty"`
	Overview     string  `json:"overview"`
	MediaType    string  `json:"media_type"` // "movie" or "tv"
	Source       string  `json:"source"`
}

// MDBListResult represents the result of fetching a list
type MDBListResult struct {
	Success     bool          `json:"success"`
	Source      string        `json:"source"`
	Movies      []MDBListItem `json:"movies"`
	Series      []MDBListItem `json:"series"`
	Total       int           `json:"total"`
	MovieCount  int           `json:"movie_count"`
	SeriesCount int           `json:"series_count"`
	FetchedAt   string        `json:"fetched_at"`
}

// MDBListConfig represents a configured list
type MDBListConfig struct {
	URL     string `json:"url"`
	Enabled bool   `json:"enabled"`
	Name    string `json:"name,omitempty"`
}

// NewMDBListClient creates a new MDBList client
func NewMDBListClient(apiKey, cacheDir string) *MDBListClient {
	if cacheDir == "" {
		cacheDir = "./cache/mdblist"
	}

	os.MkdirAll(cacheDir, 0755)

	return &MDBListClient{
		apiKey:      apiKey,
		cacheDir:    cacheDir,
		cacheExpiry: 1 * time.Hour,
		client: &http.Client{
			Timeout: 15 * time.Second,
		},
	}
}

// HasAPIKey checks if API key is configured
func (m *MDBListClient) HasAPIKey() bool {
	return m.apiKey != ""
}

// TestConnection tests the MDBList API connection
func (m *MDBListClient) TestConnection() (map[string]interface{}, error) {
	if !m.HasAPIKey() {
		return map[string]interface{}{
			"success": false,
			"error":   "API key not configured",
		}, nil
	}

	apiURL := fmt.Sprintf("https://mdblist.com/api/user/?apikey=%s", url.QueryEscape(m.apiKey))
	resp, err := m.client.Get(apiURL)
	if err != nil {
		return map[string]interface{}{
			"success": false,
			"error":   "Failed to connect to MDBList API",
		}, err
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)
	var data map[string]interface{}
	json.Unmarshal(body, &data)

	if errMsg, ok := data["error"]; ok {
		return map[string]interface{}{
			"success": false,
			"error":   errMsg,
		}, nil
	}

	return map[string]interface{}{
		"success":               true,
		"user":                  data["name"],
		"api_requests_remaining": data["api_requests"],
	}, nil
}

// GetUserLists retrieves user's MDBList lists
func (m *MDBListClient) GetUserLists() (map[string]interface{}, error) {
	if !m.HasAPIKey() {
		return map[string]interface{}{
			"success": false,
			"error":   "API key not configured",
		}, nil
	}

	apiURL := fmt.Sprintf("https://mdblist.com/api/lists/user/?apikey=%s", url.QueryEscape(m.apiKey))
	resp, err := m.client.Get(apiURL)
	if err != nil {
		return map[string]interface{}{
			"success": false,
			"error":   "Failed to fetch user lists",
		}, err
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)
	var lists []map[string]interface{}
	json.Unmarshal(body, &lists)

	// Get username from first list's details
	username := ""
	if len(lists) > 0 {
		if listID, ok := lists[0]["id"].(float64); ok {
			detailURL := fmt.Sprintf("https://mdblist.com/api/lists/%d/?apikey=%s", int(listID), url.QueryEscape(m.apiKey))
			detailResp, err := m.client.Get(detailURL)
			if err == nil {
				defer detailResp.Body.Close()
				detailBody, _ := io.ReadAll(detailResp.Body)
				var details []map[string]interface{}
				if json.Unmarshal(detailBody, &details) == nil && len(details) > 0 {
					if un, ok := details[0]["user_name"].(string); ok {
						username = un
					}
				}
			}
		}
	}

	// Add username to each list
	for i := range lists {
		lists[i]["user_name"] = username
	}

	return map[string]interface{}{
		"success":  true,
		"lists":    lists,
		"username": username,
	}, nil
}

// FetchListByURL fetches items from a public MDBList URL
func (m *MDBListClient) FetchListByURL(listURL string) (*MDBListResult, error) {
	// Parse URL to extract username and list slug
	// Format: https://mdblist.com/lists/username/listname
	var username, listSlug string
	
	if _, err := fmt.Sscanf(listURL, "https://mdblist.com/lists/%s/%s", &username, &listSlug); err == nil {
		return m.FetchPublicList(username, listSlug)
	}

	return nil, fmt.Errorf("invalid MDBList URL format")
}

// FetchPublicList fetches a public list by username and slug
func (m *MDBListClient) FetchPublicList(username, listSlug string) (*MDBListResult, error) {
	// Check cache first
	cacheKey := fmt.Sprintf("list_%s_%s", username, listSlug)
	if cached := m.getCache(cacheKey); cached != nil {
		return cached, nil
	}

	// Fetch from MDBList public JSON endpoint
	apiURL := fmt.Sprintf("https://mdblist.com/lists/%s/%s/json", username, listSlug)
	resp, err := m.client.Get(apiURL)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch list: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var items []map[string]interface{}
	if err := json.Unmarshal(body, &items); err != nil {
		return nil, fmt.Errorf("invalid JSON response: %w", err)
	}

	// Process items
	result := m.processListItems(items, fmt.Sprintf("%s/%s", username, listSlug))
	m.setCache(cacheKey, result)

	return result, nil
}

// FetchListByID fetches a list by ID (requires API key)
func (m *MDBListClient) FetchListByID(listID string) (*MDBListResult, error) {
	if !m.HasAPIKey() {
		return nil, fmt.Errorf("API key required for list ID lookup")
	}

	// Check cache
	cacheKey := fmt.Sprintf("list_id_%s", listID)
	if cached := m.getCache(cacheKey); cached != nil {
		return cached, nil
	}

	apiURL := fmt.Sprintf("https://mdblist.com/api/lists/%s/items/?apikey=%s", listID, url.QueryEscape(m.apiKey))
	resp, err := m.client.Get(apiURL)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch list: %w", err)
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)
	var items []map[string]interface{}
	json.Unmarshal(body, &items)

	result := m.processListItems(items, fmt.Sprintf("list_%s", listID))
	m.setCache(cacheKey, result)

	return result, nil
}

// SearchLists searches MDBList for lists
func (m *MDBListClient) SearchLists(query string) ([]interface{}, error) {
	apiURL := fmt.Sprintf("https://mdblist.com/api/lists/search/?query=%s", url.QueryEscape(query))
	if m.HasAPIKey() {
		apiURL += fmt.Sprintf("&apikey=%s", url.QueryEscape(m.apiKey))
	}

	resp, err := m.client.Get(apiURL)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)
	var lists []interface{}
	json.Unmarshal(body, &lists)

	return lists, nil
}

// GetTopLists gets top/popular MDBList lists
func (m *MDBListClient) GetTopLists(mediaType string) ([]interface{}, error) {
	apiURL := fmt.Sprintf("https://mdblist.com/api/lists/top/?mediatype=%s", url.QueryEscape(mediaType))
	if m.HasAPIKey() {
		apiURL += fmt.Sprintf("&apikey=%s", url.QueryEscape(m.apiKey))
	}

	resp, err := m.client.Get(apiURL)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)
	var lists []interface{}
	json.Unmarshal(body, &lists)

	return lists, nil
}

// processListItems processes raw list items into structured format
func (m *MDBListClient) processListItems(items []map[string]interface{}, sourceName string) *MDBListResult {
	var movies, series []MDBListItem

	for _, item := range items {
		tmdbID, ok := item["tmdb_id"].(float64)
		if !ok {
			if id, ok := item["id"].(float64); ok {
				tmdbID = id
			} else {
				continue
			}
		}

		mediaType := "movie"
		if mt, ok := item["mediatype"].(string); ok {
			mediaType = mt
		} else if mt, ok := item["media_type"].(string); ok {
			mediaType = mt
		}

		title := ""
		if t, ok := item["title"].(string); ok {
			title = t
		} else if n, ok := item["name"].(string); ok {
			title = n
		}

		processed := MDBListItem{
			ID:           int(tmdbID),
			TMDBID:       int(tmdbID),
			Title:        title,
			MediaType:    mediaType,
			Source:       "mdblist:" + sourceName,
		}

		if imdb, ok := item["imdb_id"].(string); ok {
			processed.IMDBID = imdb
		}
		if year, ok := item["year"].(float64); ok {
			processed.Year = int(year)
		} else if year, ok := item["release_year"].(float64); ok {
			processed.Year = int(year)
		}
		if rating, ok := item["rating"].(float64); ok {
			processed.Rating = rating
		} else if score, ok := item["score"].(float64); ok {
			processed.Rating = score
		}
		if poster, ok := item["poster"].(string); ok {
			processed.PosterPath = poster
		}
		if backdrop, ok := item["backdrop"].(string); ok {
			processed.BackdropPath = backdrop
		}
		if overview, ok := item["overview"].(string); ok {
			processed.Overview = overview
		} else if desc, ok := item["description"].(string); ok {
			processed.Overview = desc
		}

		if mediaType == "movie" {
			movies = append(movies, processed)
		} else {
			series = append(series, processed)
		}
	}

	return &MDBListResult{
		Success:     true,
		Source:      sourceName,
		Movies:      movies,
		Series:      series,
		Total:       len(movies) + len(series),
		MovieCount:  len(movies),
		SeriesCount: len(series),
		FetchedAt:   time.Now().Format("2006-01-02 15:04:05"),
	}
}

// getCache retrieves cached list data
func (m *MDBListClient) getCache(key string) *MDBListResult {
	cachePath := filepath.Join(m.cacheDir, key+".json")
	
	info, err := os.Stat(cachePath)
	if err != nil {
		return nil
	}

	// Check if cache is expired
	if time.Since(info.ModTime()) > m.cacheExpiry {
		os.Remove(cachePath)
		return nil
	}

	data, err := os.ReadFile(cachePath)
	if err != nil {
		return nil
	}

	var result MDBListResult
	if err := json.Unmarshal(data, &result); err != nil {
		return nil
	}

	return &result
}

// setCache stores list data in cache
func (m *MDBListClient) setCache(key string, result *MDBListResult) {
	cachePath := filepath.Join(m.cacheDir, key+".json")
	data, _ := json.MarshalIndent(result, "", "  ")
	os.WriteFile(cachePath, data, 0644)
}

// FetchAllConfiguredLists fetches all configured MDBList lists
func (m *MDBListClient) FetchAllConfiguredLists(configPath string) (*MDBListResult, error) {
	// Load saved lists from config
	configs, err := m.loadListConfigs(configPath)
	if err != nil {
		return &MDBListResult{
			Success: true,
			Movies:  []MDBListItem{},
			Series:  []MDBListItem{},
			Total:   0,
		}, nil
	}

	// Filter enabled lists
	var enabledLists []MDBListConfig
	for _, config := range configs {
		if config.Enabled {
			enabledLists = append(enabledLists, config)
		}
	}

	if len(enabledLists) == 0 {
		return &MDBListResult{
			Success: true,
			Movies:  []MDBListItem{},
			Series:  []MDBListItem{},
			Total:   0,
		}, nil
	}

	// Fetch all lists and merge
	allMovies := make(map[int]MDBListItem)
	allSeries := make(map[int]MDBListItem)

	for _, listConfig := range enabledLists {
		result, err := m.FetchListByURL(listConfig.URL)
		if err != nil {
			continue
		}

		// Deduplicate by TMDB ID
		for _, movie := range result.Movies {
			allMovies[movie.TMDBID] = movie
		}
		for _, show := range result.Series {
			allSeries[show.TMDBID] = show
		}
	}

	// Convert maps to slices
	movies := make([]MDBListItem, 0, len(allMovies))
	for _, m := range allMovies {
		movies = append(movies, m)
	}
	series := make([]MDBListItem, 0, len(allSeries))
	for _, s := range allSeries {
		series = append(series, s)
	}

	return &MDBListResult{
		Success:     true,
		Source:      "merged",
		Movies:      movies,
		Series:      series,
		Total:       len(movies) + len(series),
		MovieCount:  len(movies),
		SeriesCount: len(series),
		FetchedAt:   time.Now().Format("2006-01-02 15:04:05"),
	}, nil
}

// loadListConfigs loads list configurations from file
func (m *MDBListClient) loadListConfigs(configPath string) ([]MDBListConfig, error) {
	if configPath == "" {
		configPath = "./mdblist_config.json"
	}

	data, err := os.ReadFile(configPath)
	if err != nil {
		return nil, err
	}

	var configs []MDBListConfig
	if err := json.Unmarshal(data, &configs); err != nil {
		return nil, err
	}

	return configs, nil
}
